integration/install/danm-installer-config.yaml

```diff
@@ -0,0 +1,89 @@
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  namespace: kube-system
+  name: danm-installer-config
+data:
+  #
+  # DANM deployment mode. This MUST be either "lightweight" or "production". Please see the DANM user guide for
+  # details on the two modes.
+  #
+  deploy_mode: production
+
+  #
+  # CNI configuration directory. Typically, this is "/etc/cni/net.d". This is the directory where your
+  # current (bootstrap) CNI configuration is located, too.
+  #
+  cni_dir: /etc/cni/net.d
+
+  #
+  # CNI naming scheme. See section "Naming container interfaces" in the user guide for a more detailed
+  # discussion. Set this parameter to "legacy" if you wish container interface names to be set exactly
+  # according to DanmNet.Spec.Options.container_prefix, or to an empty string if you with the first
+  # interface to always be named "eth0".
+  #
+  cni_naming_scheme: ""
+
+  #
+  # [OPTIONAL] Kubernetes API Root CA certificate. If left blank, the installer will obtain the
+  # API server certificate from the Kubernetes API. Note, however, that placing a certificate here
+  # is technically more secure (as it provides external verification of the CA certificate, rather
+  # than blindly trusting the certificate that we see from the server) and also more future-proof
+  # if the individual API server's certificate ever were to change in the future, to put the Root
+  # CA certificate here. You can obtain this, for example, by running:
+  #
+  # kubectl config view \
+  #   --flatten \
+  #   -o jsonpath='{.clusters[0].cluster.certificate-authority-data} \
+  #   | base64 -d
+  #
+#  api_ca_cert: |
+#    -----BEGIN CERTIFICATE-----
+#    VGhpcyBJcyBBIFBsYWNlaG9sZGVyIFN0cmluZy4uLiBUaGlzIElzIEEgUGxhY2Vo
+#    ...
+#    VGhpcyBpcyB0aGUgZW5kIG9mIGEgcGxhY2Vob2xkZXIgc3RyaW5nCg==
+#    -----END CERTIFICATE-----
+
+  #
+  # This is the type of the CNI plugin used for your default (bootstrap) network.
+  # The value can also be found in the "type" field of your bootstrap CNI configuration
+  # file, eg. "cat ${cni_dir}/${default_cni_network_id}.conf | jq -Mr '.type'"
+  #
+  default_cni_type: calico
```

Comment:
does this need to be a manually provided parameter? I think it could be grepped from the default CNI config, decreasing the chance of manual misconfiguration happening

Reply:
Depends on the default CNI. Calico has a kubeconfig that we could fetch this from; Flannel, on the other hand, does not.

We can (and do, in my proposed code) fetch the API server certificate from the API if no certificate is provided. So we can do without it -- but technically, the API server certificate is not the same as the Root CA certificate. I'm not aware of any good way to fetch the Root CA certificate programmatically, "from the inside".

Reply:
I didn't mean the CA, I only meant the name of the CNI binary which needs to be invoked for this network. I guess that is the only thing this parameter would contorl right?
and that is a mandatory attribute in a CNI config file, so I thought we could just grep it

Reply:
Err yes, sorry, I should have learned to read. You're right, we can probably fetch this from the CNI config file... I'll look into that.

Reply:
So this isn't completely trivial, but certainly doable.

The reason it is not completely trivial, is because the installer job is given the CNI configuration file directory through a configmap (`/etc/cni/net.d`), which it then uses to construct the volume mounts for the DaemonSet -- but the installer job itself, does not have `/etc/cni/net.d` mounted. For the installer to mount this, it would mean that the installer's own podspec also needs to have that volume attached. That is easy to do, of course, but would require that the user has to provide the CNI configuration path twice -- once inside the configMap, and once inside the installer podspec. This means we're trading one bit of extra work (providing the cni type) against another one (duplicating the mount directory).
I guess this is one of the places where Helm would help.

Another option is that the installer job could query this from the daemonsets. That is, the daemonsets (which have the directory mounted) could expose that information, either by a log message or by an API of sorts -- where the API could be, in the most basic, most crude scenario, the installer doing a `kubectl exec cat /host/cni/net.d/${filename}` on one of the daemonset pods. This should be fairly easy to implement, but requires some sequencing -- the installer needs to wait for the DS to be alive, *then* fetch the CNI type from one member, and *then* render the default network config. Still, this would probably be the simplest solution that does not require any additional input or duplication from the user.


As a side note, I also pushed another change right now that allows the DANM installer to *write* the boostrap CNI configuration file, rather than assuming it pre-exists. (I found that in pretty much all my Kubernetes deployments, the default CNI configuration that was installed when the cluster is created, is a .conflist - so I find myself having to deploy a different config time after time.) That, of course, solves the problem in its own way: If the installer knows the bootstrap CNI config because it is creating it, then we could simply glean the type from the config we were given. But IMHO this feature (DANM installer writing the bootstrap CNI config) should remain optional, not required by design....

Reply:
yep, you are right. totally does not worth the effort :)